### PR TITLE
*: backup to S3 and restore from S3 #116

### LINKED ISSUE #116: [feature]add xtrabackup manger
**Is your feature request related to a problem? Please describe.**

we need xtrabackup to cloud for the cluster
**Describe the solution you'd like**
Make a backup like this:
```shell
kubectl apply -f mysql_v1alpha1_backup.yaml 
```
To backup the database data to S3 storage , or Persisten volume
Then could use the backup copy to create cluster , add RestoreFrom feild in mysql_v1alpha1_cluster.yaml
```shell
kubectl apply -f mysql_v1alpha1_cluster.yaml 
```
create the cluster from backup copy

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by acekingke**
### What type of PR is this?

/enhancement
### Which issue(s) this PR fixes?

Fixes #116 

### What this PR does?

Summary:

### CODE REVIEW

**Comment by runkecheng** [comment]
Need to improve the use document #151

File: Dockerfile.sidecar [hunk 1]
```diff
@@ -24,12 +25,32 @@ RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o bin/sidecar cmd/sidecar
...
+    --home-dir /var/lib/mysql \
+    --no-create-home \
+    --gid mysql \
+    mysql;
 
-# Create a group and user
-RUN addgroup -g 1001 mysql && adduser -u 1001 -g 1001 -S mysql
+ENV PS_VERSION 5.7.33-36-1
+ENV OS_VER focal
+ENV FULL_PERCONA_VERSION "$PS_VERSION.$OS_VER"
```
Comment (andyli029) [inline Dockerfile.sidecar:42]: Delete the line.

**Comment by andyli029** [comment]
#144

File: cmd/sidecar/main.go [hunk 2]
```diff
@@ -49,12 +50,61 @@ var (
...
-
 	cfg := sidecar.NewConfig()
-
+	stop := make(chan struct{})
 	initCmd := sidecar.NewInitCommand(cfg)
 	cmd.AddCommand(initCmd)
 
+	takeBackupCmd := &cobra.Command{
+		Use:   "backup",
+		Short: "Take a backup from node and push it to rclone path.",
```
Comment (andyli029) [inline cmd/sidecar/main.go:60]: Fix the commets about `rclone`

File: charts/mysql-operator/crds/mysql.radondb.com_backups.yaml [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            description: BackupSpec defines the desired state of Backup
+            properties:
+              clustname:
+                description: Cluster represents the cluster name to backup
+                type: string
+              hostname:
+                description: HostName represents the host for which to take backup
+                type: string
+              image:
+                default: acekingke/sidecar:0.1
```
Comment (andyli029) [inline charts/mysql-operator/crds/mysql.radondb.com_backups.yaml:46]: Fix the image

File: backup/syncer/job.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: sctName,
+					},
+					Key:      "backup-user",
+					Optional: &optTrue,
+				},
+			},
+		},
+		//backup user  for sidecar http server
```
Comment (andyli029) [inline backup/syncer/job.go:154]: Fix the comment.

File: controllers/backup_controller.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+// Update backup Object and Status
+func (r *BackupReconciler) updateBackup(savedBackup *apiv1alpha1.Backup, backup *backup.Backup) error {
+	log := log.Log.WithName("controllers").WithName("Backup")
+	if !reflect.DeepEqual(savedBackup, backup.Unwrap()) {
+		if err := r.Update(context.TODO(), backup.Unwrap()); err != nil {
+			return err
+		}
+	}
+	if !reflect.DeepEqual(savedBackup.Status, backup.Unwrap().Status) {
+
```
Comment (andyli029) [inline controllers/backup_controller.go:143]: Delete the space line.

File: charts/mysql-operator/values.yaml [hunk 6]
```diff
@@ -15,7 +15,7 @@ fullnameOverride: ""
 
 manager:
   image: radondb/mysql-operator
-  tag: 0.1
+  tag: 0.1.88
```
Comment (runkecheng) [inline charts/mysql-operator/values.yaml:18]: if use ci, it should be `latest`

File: config/samples/mysql_v1alpha1_cluster.yaml [hunk 7]
```diff
@@ -55,7 +62,7 @@ spec:
 
   podSpec:
     imagePullPolicy: IfNotPresent
-    sidecarImage: radondb/mysql-sidecar:0.1
+    sidecarImage: radondb/mysql-sidecar:0.1.88
```
Comment (runkecheng) [inline config/samples/mysql_v1alpha1_cluster.yaml:65]: if use ci, it should be latest

File: controllers/backup_controller.go [hunk 8]
```diff
@@ -114,3 +150,13 @@ func (r *BackupReconciler) updateBackup(savedBackup *apiv1alpha1.Backup, backup
 	}
 	return nil
 }
+
+// Check the job is finished.
+func IsJobFinished(job *batchv1.Job) bool {
```
Comment (andyli029) [inline controllers/backup_controller.go:154]: Reference: 
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/utils.go

File: controllers/backup_controller.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/presslabs/controller-util/syncer"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (andyli029) [inline controllers/backup_controller.go:27]: corev1

File: controllers/backup_controller.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+		Namespace: req.Namespace}); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	var finishedBackups []*batchv1.Job
+	for _, job := range backups.Items {
+		if IsJobFinished(&job) {
+			finishedBackups = append(finishedBackups, &job)
+		}
+
```
Comment (andyli029) [inline controllers/backup_controller.go:104]: Delete the space line.

File: controllers/backup_controller.go [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+	if err := syncer.Sync(ctx, jobSyncer, r.Recorder); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err = r.updateBackup(savedBackup, backup); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Clear the backup, Just keep historyLimit len
+	backups := batchv1.JobList{}
```
Comment (andyli029) [inline controllers/backup_controller.go:93]: Merge `93-123` into one function

**Comment by runkecheng** [comment]
The init-mysql and backup containers use the same image, the configuration items checked at startup are the same.  but the required configuration items are different. So some unnecessary logs will be generated like:
```
2021-09-01T09:01:47.203Z        INFO    sidecar environment is not set  {"key": "MYSQL_VERSION"}
```

File: config/samples/mysql_v1alpha1_backup.yaml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+apiVersion: mysql.radondb.com/v1alpha1
+kind: Backup
+metadata:
+  name: backup-sample1
```
Comment (runkecheng) [inline config/samples/mysql_v1alpha1_backup.yaml:4]: remove 1

File: utils/constants.go [hunk 13]
```diff
@@ -37,12 +37,13 @@ const (
...
-	ContainerSlowLogName  = "slowlog"
-	ContainerAuditLogName = "auditlog"
-	ContainerBackupName   = "backup"
+	ContainerMysqlName     = "mysql"
+	ContainerXenonName     = "xenon"
+	ContainerMetricsName   = "metrics"
+	ContainerSlowLogName   = "slowlog"
+	ContainerAuditLogName  = "auditlog"
+	ContainerBackupName    = "backup"
+	ContainerBackupJobName = "job-backup"
```
Comment (andyli029) [inline utils/constants.go:46]: Change the name.

File: docs/DEPLOY.md [hunk 14]
```diff
@@ -0,0 +1,127 @@
+# mysql-operator
```
Comment (andyli029) [inline docs/DEPLOY.md:1]: Add backup_restore_s3 to name of the file

**Pull request merged**